PubnativeLite/PubnativeLiteDemo/Adapters/AppMonet Adapters/MoPub Adapters/AppMonet+MoPub.m

```diff
@@ -23,45 +23,73 @@
 #import "AppMonet+MoPub.h"
 #import <HyBid/HyBid.h>
 
-@interface AppMonet ()
+@interface AppMonet () <HyBidAdRequestDelegate>
 
-typedef struct { MPAdView *adView; void (^onReadyBlock)(void); } AppMonetAdStruct;
-typedef struct { MPInterstitialAdController *interstitial; void (^onReadyBlock)(void); } AppMonetInterstitialAdStruct;
-
-@property (class, nonatomic) NSMutableDictionary *dict;
-@property (class, nonatomic) NSMutableDictionary *interstitialDict;
-@property (class, nonatomic, strong) HyBidAdRequest *adRequest;
-@property (class, nonatomic, strong) HyBidInterstitialAdRequest *interstitialAdRequest;
+typedef struct { MPAdView *adView;} AppMonetAdStruct;
+typedef struct { MPInterstitialAdController *interstitial;} AppMonetInterstitialAdStruct;
 
 @end
 
 @implementation AppMonet (MoPub)
 
+static void (^_onReadyBlock)(void);
+
+static NSMutableDictionary *_dict=nil;
++ (NSMutableDictionary *)dict {
+  if (_dict == nil) {
+    _dict = [[NSMutableDictionary alloc] init];
+  }
+  return _dict;
+}
+
+static NSMutableDictionary *_interstitialDict=nil;
++ (NSMutableDictionary *)interstitialDict {
+  if (_interstitialDict == nil) {
+      _interstitialDict = [[NSMutableDictionary alloc] init];
+  }
+  return _interstitialDict;
+}
+
+static HyBidAdRequest *_adRequest=nil;
++ (HyBidAdRequest *)adRequest {
+  if (_adRequest == nil) {
+      _adRequest = [[HyBidAdRequest alloc] init];
+  }
+  return _adRequest;
+}
+
+static HyBidInterstitialAdRequest *_interstitialAdRequest=nil;
++ (HyBidInterstitialAdRequest *)interstitialAdRequest {
+  if (_interstitialAdRequest == nil) {
+      _interstitialAdRequest = [[HyBidInterstitialAdRequest alloc] init];
+  }
+  return _interstitialAdRequest;
+}
+
 + (void)addBids:(MPAdView *)adView andAppMonetAdUnitId:(NSString *)appMonetAdUinitId andTimeout:(NSNumber *)timeout :(void (^)(void))onReadyBlock
 {
     AppMonetAdStruct appMonetStruct;
     appMonetStruct.adView = adView;
-    appMonetStruct.onReadyBlock = onReadyBlock;
```

Comment:
but why did you make `onReadyBlock` as a separate parameter and removed it from the struct? If there will be 2 ad requests from one screen let's say this will cause problems, as it will be overwritten

Reply:
from the struct i was not able to get the completion block called. I got an error BAD_Access even after modifying the struct to be passed as NSData

Reply:
@faresbh then we have to find a solution to pass that error. Otherwise we can get concurrency issues

Reply:
Maybe we can make it a class or something? But we do need to separate them otherwise it'll be impossible to have simultaneous ad requests

Reply:
yes you're right about simultaneous ad requests. I will try to make it as a class.